contracts/psm/PSM.sol

```diff
@@ -0,0 +1,293 @@
+    event SetFeeReceiver(address feeReceiver);
+    event BuyToken(address account, uint256 realAmount, uint256 fee);
+    event SellToken(address account, uint256 realAmount, uint256 fee);
+    event SetDailyLimit(uint256 dailyLimit);
+    event SetMinSell(uint256 minSell);
+    event SetMinBuy(uint256 minBuy);
+    event SetVaultManager(address vaultManager);
+    event EmergencyWithdraw(address token, uint256 amount);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+      * @dev initialize contract
+      * @param _admin admin address
+      * @param _manager manager address
+      * @param _token token address
+      * @param _feeReceiver fee receiver address
+      * @param _lisUSD lisUSD address
+      * @param _sellFee sell fee
+      * @param _buyFee buy fee
+      * @param _minSell min sell amount
+      * @param _minBuy min buy amount
+      */
+    function initialize(
+        address _admin,
+        address _manager,
+        address _token,
+        address _feeReceiver,
+        address _lisUSD,
+        uint256 _sellFee,
+        uint256 _buyFee,
+        uint256 _dailyLimit,
+        uint256 _minSell,
+        uint256 _minBuy
+    ) public initializer {
+        require(_admin != address(0), "admin cannot be zero address");
+        require(_manager != address(0), "manager cannot be zero address");
+        require(_token != address(0), "token cannot be zero address");
+        require(_feeReceiver != address(0), "feeReceiver cannot be zero address");
+        require(_lisUSD != address(0), "lisUSD cannot be zero address");
+        require(_sellFee <= FEE_PRECISION, "sellFee must be less or equal than FEE_PRECISION");
+        require(_buyFee <= FEE_PRECISION, "buyFee must be less or equal than FEE_PRECISION");
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+        __Pausable_init();
+        __UUPSUpgradeable_init();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+        _setupRole(MANAGER, _manager);
+
+        token = _token;
+        sellFee = _sellFee;
+        buyFee = _buyFee;
+        feeReceiver = _feeReceiver;
+        lisUSD = _lisUSD;
+
+        dailyLimit = _dailyLimit;
+        minSell = _minSell;
+        minBuy = _minBuy;
+
+        emit SetBuyFee(_buyFee);
+        emit SetSellFee(_sellFee);
+        emit SetFeeReceiver(_feeReceiver);
+        emit SetDailyLimit(_dailyLimit);
+        emit SetMinSell(_minSell);
+        emit SetMinBuy(_minBuy);
+    }
+
+    /**
+     * @dev sell token to get lisUSD
+     * @param amount token amount
+     */
+    function sell(uint256 amount) external nonReentrant whenNotPaused {
+        // calculate fee and real amount
+        uint256 fee = Math.mulDiv(amount, sellFee, FEE_PRECISION);
+        uint256 realAmount = amount - fee;
+
+        // check sell limit
+        require(realAmount <= IERC20(lisUSD).balanceOf(address(this)), "exceed sell limit");
```

Comment:
This should be original amount, otherwise, the line fee transfer action will fail.